src/utils/memoization.js

```diff
@@ -0,0 +1,289 @@
+ */
+export function compareLists(listA, listB, comparator = Object.is) {
+  if (listA === listB) {
+    return true
+  }
+
+  if (listA.length !== listB.length) {
+    return false
+  }
+
+  for (let i = 0; i < listA.length; i++) {
+    if (!comparator(listA[i], listB[i])) {
+      return false
+    }
+  }
+
+  return true
+}
+
+/**
+ * Compare two lists, ignoring the order of the items.
+ * @template T
+ * @param {T[]} listA
+ * @param {T[]} listB
+ * @param {(itemA: T, itemB: T) => boolean} matcher
+ *   a function to see if two items represent the same thing. This should be a quick check
+ *   since this gets called for essentially every pair of items. If this is true, then the
+ *   more complex comparator function will be called.
+ * @param {(itemA: T, itemB: T) => boolean} comparator
+ *   a function to see if two items are equal.
+ */
+export function compareUnorderedLists(
+  listA,
+  listB,
+  matcher = Object.is,
+  comparator = Object.is
+) {
+  if (listA === listB) {
+    return true
+  }
+
+  if (listA.length !== listB.length) {
+    return false
+  }
+
+  const matchedIndexes = new Set()
+  for (const itemA of listA) {
+    const indexB = listB.findIndex(
+      (item, index) => !matchedIndexes.has(index) && matcher(itemA, item)
+    )
+    if (indexB === -1) {
+      return false
+    }
+
+    matchedIndexes.add(indexB)
+    const itemB = listB[indexB]
+
+    if (!comparator(itemA, itemB)) {
+      return false
+    }
+  }
+
+  return true
+}
+
+/**
+ * @template [T=Record<string, unknown>]
+ * @param {T} objectA
+ * @param {T} objectB
+ * @param {Object} options
+ *   optional. If excluded, the logic here should match React's default memoization logic.
+ * @param {Partial<Record<keyof T, (valueA: any, valueB: any) => boolean>>=} options.comparators
+ *   an object mapping keys of the objects to a comparator override.
+ *   Useful if some keys are simple checks like strings/numbers but other keys need unique comparators
+ * @param {(valueA: any, valueB: any) => boolean=} options.defaultComparator
+ *   the fallback comparator if not explicitly defined in `comparators`.
+ *   Defaults to `Object.is` to be consistent with React's default memoization logic
+ * @param {(keyof T)[]=} options.ignoredKeys
+ *   these keys will be ignored when comparing objects
+ * @param {Boolean=} options.logDifferences
+ *   a boolean to indicate if debug messages should be logged when objects change
+ * @param {string=} options.from
+ *   a string to help identify where a change is triggered from when debug is on
+ */
+export function compareObjects(
+  objectA,
+  objectB,
+  {
+    comparators = {},
+    defaultComparator = Object.is,
+    ignoredKeys = [],
+    logDifferences,
+    from,
+  } = {}
+) {
+  if (objectA === objectB) {
+    return true
+  }
+
+  // If the number of checked keys has changed, we don't need to check any of the values.
+  const ignoredKeysSet = new Set(ignoredKeys)
+  const keysToCheckA = Object.keys(objectA).filter(
```

Comment:
if conferenceData is null/undefined (swing from null/undefined), wouldn't `Object.keys` cause a `TypeError`?